### PR TITLE
feature: save, add, remove

### PR DESCRIPTION
**Pull request by Kupov**

### CODE REVIEW

**Comment by Kupov** [comment]
@dontuse

File: src/Table/actions.js [hunk 1]
```diff
@@ -96,14 +100,34 @@ export const endDrag = payload => ({
...
+
 export const focusNext = payload =>
   (dispatch, getState) => {
     const state = getState();
     dispatch({
       type: TABLE_EDITOR_CELL_FOCUS_NEXT,
       payload: {
         ...payload,
-        rows: state.rows.present
+        rows: state.history.current
```
Comment (apress-hound) [inline src/Table/actions.js]: Expected ')' and instead saw ':'.<br>Missing semicolon.

File: src/SaveControl/reducer.js [hunk 2]
```diff
@@ -1,49 +1,320 @@
...
+  return transformForServer([...diff, ...deletedItems, ...newItems]);
 };
 
 export default function (state = initialState, action) {
   switch (action.type) {
     case TABLE_EDITOR_LOAD_SUCCESS:
       return {
         ...state,
-        prevProps: action.payload.rows
+        prevState: action.payload.rows
```
Comment (apress-hound) [inline src/SaveControl/reducer.js]: Missing semicolon.<br>Expected 'case' and instead saw ':'.<br>Expected '}' to match '{' from line 175 and instead saw ':'.<br>Unrecoverable syntax error. (55% scanned).

File: src/Error/reducer.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  switch (action.type) {
+    case ERROR_ADD:
+      return [
+        ...state,
+        {
+          id: gerErrorId(),
+          type: action.payload.type,
+          target: action.payload.target,
+          title: action.payload.title,
+          row_id: action.payload.row_id,
```
Comment (apress-hound) [inline src/Error/reducer.js]: Identifier 'row_id' is not in camel case.

File: src/Error/reducer.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import {ERROR_ADD, ERROR_REMOVE} from './actions';
+
+let errorId = 0;
+
+const gerErrorId = () => (errorId += 1);
+
+export default function (state = [], action) {
```
Comment (apress-hound) [inline src/Error/reducer.js]: Regular parameters should not come after default parameters.

File: src/Error/e-error.scss [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  &-wrapper { position: fixed; top: 32px; right: 0; z-index: 9999; }
+  &-repeat {
+    text-decoration: underline; cursor: pointer; color: #fff;
+    &:link,
+    &:active,
+    &:hover { color: #fff; }
+  }
+  &-close {
+    @include close-icon(#fa6660); cursor: pointer; position: absolute; top: 8px; right: 8px; width: 16px; height: 16px; background-color: #fff; border-radius: 50%;
+    &:hover { opacity: 0.7; }
```
Comment (apress-hound) [inline src/Error/e-error.scss]: `0.7` should be written without a leading zero as `.7`

File: dist/assets-help.bundle.js [hunk 6]
```diff
@@ -78,7 +78,7 @@ return /******/ (function(modules) { // webpackBootstrap
 /******/ 	}
 /******/ 	
 /******/ 	var hotApplyOnUpdate = true;
-/******/ 	var hotCurrentHash = "e66c1f7653f01295a524"; // eslint-disable-line no-unused-vars
+/******/ 	var hotCurrentHash = "6d22d395e275000d773d"; // eslint-disable-line no-unused-vars
```
Comment (apress-hound) [inline dist/assets-help.bundle.js]: 'hotCurrentHash' was used before it was defined.<br>Mixed double and single quotes.

File: src/Table/reducers/history.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  current: []
+};
+
+export default function history(state = initialState, action) {
+  switch (action.type) {
+    case TABLE_EDITOR_LOAD_SUCCESS:
+      return {
+        ...state,
+        current: rows(state.current, action),
+        next: [],
```
Comment (apress-hound) [inline src/Table/reducers/history.js]: Missing semicolon.<br>Unrecoverable syntax error. (21% scanned).

File: src/Table/reducers/history.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  next: [],
+  current: []
+};
+
+export default function history(state = initialState, action) {
+  switch (action.type) {
+    case TABLE_EDITOR_LOAD_SUCCESS:
+      return {
+        ...state,
+        current: rows(state.current, action),
```
Comment (apress-hound) [inline src/Table/reducers/history.js]: Missing semicolon.<br>Expected 'case' and instead saw ':'.<br>Expected '}' to match '{' from line 19 and instead saw ':'.

File: src/Table/reducers/history.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  prev: [],
+  next: [],
+  current: []
+};
+
+export default function history(state = initialState, action) {
+  switch (action.type) {
+    case TABLE_EDITOR_LOAD_SUCCESS:
+      return {
+        ...state,
```
Comment (apress-hound) [inline src/Table/reducers/history.js]: Expected '}' to match '{' from line 22 and instead saw '...'.<br>Missing semicolon.

File: src/Table/reducers/history.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+  HISTORY_PREV
+} from '../actions';
+
+const initialState = {
+  prev: [],
+  next: [],
+  current: []
+};
+
+export default function history(state = initialState, action) {
```
Comment (apress-hound) [inline src/Table/reducers/history.js]: Regular parameters should not come after default parameters.

File: src/SaveControl/reducer.js [hunk 11]
```diff
@@ -1,49 +1,328 @@
...
```
Comment (apress-hound) [inline src/SaveControl/reducer.js]: Missing semicolon.<br>Expected 'case' and instead saw ':'.<br>Expected '}' to match '{' from line 175 and instead saw ':'.<br>Unrecoverable syntax error. (54% scanned).

File: src/Table/reducers/history.js (hunk 7 above)
Comment (apress-hound) [inline src/Table/reducers/history.js]: Missing semicolon.<br>Unrecoverable syntax error. (21% scanned).

File: src/Table/reducers/history.js (hunk 8 above)
Comment (apress-hound) [inline src/Table/reducers/history.js]: Missing semicolon.<br>Expected 'case' and instead saw ':'.<br>Expected '}' to match '{' from line 19 and instead saw ':'.

File: src/Table/reducers/history.js (hunk 9 above)
Comment (apress-hound) [inline src/Table/reducers/history.js]: Expected '}' to match '{' from line 22 and instead saw '...'.<br>Missing semicolon.

File: src/Table/reducers/history.js (hunk 10 above)
Comment (apress-hound) [inline src/Table/reducers/history.js]: Regular parameters should not come after default parameters.

File: src/Table/actions.js (hunk 1 above)
Comment (apress-hound) [inline src/Table/actions.js]: Expected ')' and instead saw ':'.<br>Missing semicolon.

File: src/SaveControl/SaveControl.jsx [hunk 12]
```diff
@@ -6,72 +6,25 @@ import './e-save-control.scss';
...
-            diff[key] = cellDiff;
-          }
-        } else {
-          diff[key] = curStateObj[key];
-        }
-      }
-    });
 
-    return diff;
+    if (nextProps.save.isSave && nextProps.save.isProgress) {
```
Comment (Kupov) [inline src/SaveControl/SaveControl.jsx]: тут все ок, ток не сохранил правку одну

File: src/Filter/Filter.jsx [hunk 13]
```diff
@@ -3,23 +3,21 @@ import {browserHistory} from 'react-router';
```
Comment (Kupov) [inline src/Filter/Filter.jsx]: есть идеи, так скажи

File: src/SaveControl/reducer.js [hunk 14]
```diff
@@ -1,49 +1,328 @@
...
-  error: [],
-  prevProps: null
+  prevState: [],
+  saveProgress: false,
+  waitingState: [],
+  saveState: []
+};
+
+
+const getFieldId = () => 'check';
```
Comment (Kupov) [inline src/SaveControl/reducer.js]: смысл в том чтоб вынести из дифа

File: src/SaveControl/reducer.js [hunk 15]
```diff
@@ -1,49 +1,328 @@
...
+
+const getDiff = (cur, prev) => {
+  if (Array.isArray(prev) && prev.length) {
+    const tmpDiff = [];
+    prev.forEach((prevChild) => {
+      const curChild = cur.find(find.bind({}, prevChild));
+      if (curChild) {
+        const childDiff = getDiff(curChild, prevChild);
+
+        if (childDiff && typeof childDiff === 'object' && Object.keys(childDiff).length) {
```
Comment (Kupov) [inline src/SaveControl/reducer.js]: пока да, дальше посмотрим

File: src/SaveControl/reducer.js [hunk 16]
```diff
@@ -1,49 +1,328 @@
...
+        } else {
+          newObj[key] = record[key];
+        }
+      }
+    });
+
+    return newObj;
+  });
+
+const validation = (column, row) => {
```
Comment (Kupov) [inline src/SaveControl/reducer.js]: ну тут вроде и так понятно что валидирует

**Comment by Kupov** [comment]
:up:

File: src/SaveControl/reducer.js (hunk 15 above)
Comment (Kupov) [inline src/SaveControl/reducer.js]: сломается только если массив в массив будет вложен

File: src/Error/Error.jsx [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+      </div>
+    );
+  }
+}
+
+const mapStateToProps = state => ({
+  errors: state.error,
+});
+
+export default connect(mapStateToProps)(Error);
```
Comment (dontuse) [inline src/Error/Error.jsx]: Error

**Pull request merged**